synapse/storage/databases/state/bg_updates.py

```diff
@@ -98,153 +98,100 @@ def _get_state_groups_from_groups_txn(
+        # if there are too many individual conditions. 10 is an arbitrary number
+        # with no testing behind it but we do know that we specifically made this
+        # optimization for when we grab the necessary state out for
+        # `filter_events_for_client` which just uses 2 conditions
+        # (`EventTypes.RoomHistoryVisibility` and `EventTypes.Member`).
+        if use_condition_optimization and len(state_filter_condition_combos) < 10:
+            select_clause_list: List[str] = []
+            for etype, skey in state_filter_condition_combos:
+                if skey is None:
+                    where_clause = "(type = ?)"
+                    overall_select_query_args.extend([etype])
+                else:
+                    where_clause = "(type = ? AND state_key = ?)"
+                    overall_select_query_args.extend([etype, skey])
+
+                select_clause_list.append(
+                    f"""
+                    (
+                        SELECT DISTINCT ON (type, state_key)
+                            type, state_key, event_id
+                        FROM state_groups_state
+                        INNER JOIN sgs USING (state_group)
+                        WHERE {where_clause}
+                        ORDER BY type, state_key, state_group DESC
                     )
+                    """
+                )
 
-                overall_select_clause = " UNION ".join(select_clause_list)
-            else:
-                where_clause, where_args = state_filter.make_sql_filter_clause()
-                # Unless the filter clause is empty, we're going to append it after an
-                # existing where clause
-                if where_clause:
-                    where_clause = " AND (%s)" % (where_clause,)
-
-                overall_select_query_args.extend(where_args)
-
-                overall_select_clause = f"""
-                    SELECT DISTINCT ON (type, state_key)
-                        type, state_key, event_id
-                    FROM state_groups_state
-                    WHERE state_group IN (
-                        SELECT state_group FROM sgs
-                    ) {where_clause}
-                    ORDER BY type, state_key, state_group DESC
-                """
-
-            for group in groups:
-                args: List[Union[int, str]] = [group]
-                args.extend(overall_select_query_args)
-
-                txn.execute(sql % (overall_select_clause,), args)
-                for row in txn:
-                    typ, state_key, event_id = row
-                    key = (intern_string(typ), intern_string(state_key))
-                    results[group][key] = event_id
+            overall_select_clause = " UNION ".join(select_clause_list)
         else:
-            max_entries_returned = state_filter.max_entries_returned()
-
             where_clause, where_args = state_filter.make_sql_filter_clause()
```

Comment:
> Is this worth resurrecting?

@DMRobertson It seems possible to accomplish but there are a few more hurdles. If you're keen to take it over, feel free. I won't be looking at it until I get more backend facing again.